# Technical Test Phase II: 
    Author: David Horacio Duarte
    Email: dhduartea@gmail.com
![OTF](https://media.licdn.com/dms/image/D4D0BAQGF5wSwltQCRg/company-logo_200_200/0/1684924515817/onthefuze_logo?e=1716422400&v=beta&t=0sVCpKq8OAnziW10yV8vkETh2ui5D-66qDRyCWEjTSo)

Welcome to my technical test for Data Engineer & Analyst position. Throughout this test I will try to demonstrate my experience in Python programming and integrate it with the HubSpot API.  I sincerely appreciate the opportunity and look forward to continue with this process. The process will be broken down into the following sections:
1. Extracting the data from the HubSpot API
2. Transforming the data.
3. Managing duplicates.
4. Graphing the data.
5. Importing the data into HubsSpot.


## Libraries

- First import the required libraries and modules for the project.

In [1]:
import pandas as pd
import numpy as np
import requests
import re
import simplejson as json
import plotly.express as px
import plotly.graph_objects as go
import Library as lib # Personal library of functions

- Set the API keys

In [2]:
api_key_hubspot="pat-na1-3c7b0af9-bb66-40e7-a256-ce4c5eb27e81"
api_key_david="pat-na1-87c92179-b31b-43bf-bf42-f18beda31b79"


## 1. Data Extraction

In this section, we will be using the `contactExtraction` function to extract the contact information from the Hubspot API. 

In [3]:
df=lib.contactExtraction(api_key_hubspot)
df

address    country hs_object_id  \
0             Blackpool  Rue, 6576  Waterford       416102   
1          Parkfield  Avenue, 5340    Ireland       413403   
2              Abourne   Lane, 876    Ireland       417951   
3     Chester      Crossroad, 7070     Dublin       419852   
4        Tilloch   Crossroad, 8332     Dublin       425352   
...                            ...        ...          ...   
6931             Cato   Hill, 5704  Waterford       463801   
6932        Howard Crossroad, 9998     London       454595   
6933              Walnut Rue, 3859     Dublin       454631   
6934         Kimberley  Pass, 3455   Plymouth       457622   
6935        Yorkshire  Alley, 5843       Cork       461395   

                  industry          phone  \
0         Poultry and fish  0-774-386-624   
1     Fruit and vegetables  6-777-367-783   
2                  Milling  5-618-556-540   
3           Dairy products  1-161-604-327   
4                     Meat  5-645-416-200   
...                    ...            ...   
6931  Fruit and vegetables  2-301-310-740   
6932       Bakery products  6-747-016-018   
6933          Animal feeds  5-516-171-174   
6934       Bakery products  0-605-727-343   
6935  Fruit and vegetables  3-814-518-751   

                                              raw_email  \
0       Zoe <zoe_owen450104633@acrit.org> Contact Info.   
1     Zara <zara_rodwell1398442854@nickia.com> Conta...   
2     Zara <zara_freeburn1593147546@gmail.com> Conta...   
3     Winnie <winnie_walter538064895@sheye.org> Cont...   
4     Zoe <zoe_owen1652446013@bungar.biz> Contact Info.   
...                                                 ...   
6931  Rhea <rhea_grady231044837@deons.tech> Contact ...   
6932  Hailey <hailey_farrell934796609@brety.org> Con...   
6933  Caleb <caleb_purvis1251615808@iatim.tech> Cont...   
6934  Gina <gina_weasley83573127@naiker.biz> Contact...   
6935  Aeris <aeris_walsh769266811@gompie.com> Contac...   

     technical_test___create_date  
0                      2021-07-13  
1                      2021-01-09  
2                      2021-08-30  
3                      2021-02-10  
4                      2021-11-02  
...                           ...  
6931                   2021-04-24  
6932                   2021-12-04  
6933                   2021-02-23  
6934                   2021-10-02  
6935                   2021-05-06  

[6936 rows x 7 columns]

## 2. Data Transformation
For this part, we will use the dataframe extracted in the previous part. We will transform the data with some functions to get the desired output.


### 2.1. Country and City Recognition
We will use the function `recognition` in the module `countryRecognition` to recognize the country and the city of each contact and also, return a column with the tuple with the country and the city.

In [4]:
df=lib.recognition(df)
df

address  country hs_object_id  \
0             Blackpool  Rue, 6576  Ireland       416102   
1          Parkfield  Avenue, 5340  Ireland       413403   
2              Abourne   Lane, 876  Ireland       417951   
3     Chester      Crossroad, 7070  Ireland       419852   
4        Tilloch   Crossroad, 8332  Ireland       425352   
...                            ...      ...          ...   
6931             Cato   Hill, 5704  Ireland       463801   
6932        Howard Crossroad, 9998  England       454595   
6933              Walnut Rue, 3859  Ireland       454631   
6934         Kimberley  Pass, 3455  England       457622   
6935        Yorkshire  Alley, 5843  Ireland       461395   

                  industry          phone  \
0         Poultry and fish  0-774-386-624   
1     Fruit and vegetables  6-777-367-783   
2                  Milling  5-618-556-540   
3           Dairy products  1-161-604-327   
4                     Meat  5-645-416-200   
...                    ...            ...   
6931  Fruit and vegetables  2-301-310-740   
6932       Bakery products  6-747-016-018   
6933          Animal feeds  5-516-171-174   
6934       Bakery products  0-605-727-343   
6935  Fruit and vegetables  3-814-518-751   

                                              raw_email  \
0       Zoe <zoe_owen450104633@acrit.org> Contact Info.   
1     Zara <zara_rodwell1398442854@nickia.com> Conta...   
2     Zara <zara_freeburn1593147546@gmail.com> Conta...   
3     Winnie <winnie_walter538064895@sheye.org> Cont...   
4     Zoe <zoe_owen1652446013@bungar.biz> Contact Info.   
...                                                 ...   
6931  Rhea <rhea_grady231044837@deons.tech> Contact ...   
6932  Hailey <hailey_farrell934796609@brety.org> Con...   
6933  Caleb <caleb_purvis1251615808@iatim.tech> Cont...   
6934  Gina <gina_weasley83573127@naiker.biz> Contact...   
6935  Aeris <aeris_walsh769266811@gompie.com> Contac...   

     technical_test___create_date       city          countryTuple  
0                      2021-07-13  Waterford  (Ireland, Waterford)  
1                      2021-01-09        NaN        (Ireland, nan)  
2                      2021-08-30        NaN        (Ireland, nan)  
3                      2021-02-10     Dublin     (Ireland, Dublin)  
4                      2021-11-02     Dublin     (Ireland, Dublin)  
...                           ...        ...                   ...  
6931                   2021-04-24  Waterford  (Ireland, Waterford)  
6932                   2021-12-04     London     (England, London)  
6933                   2021-02-23     Dublin     (Ireland, Dublin)  
6934                   2021-10-02   Plymouth   (England, Plymouth)  
6935                   2021-05-06       Cork       (Ireland, Cork)  

[6936 rows x 9 columns]

#### 2.2. Found Email
In the following section, we will be using the data from the previous section to extract the email from raw_email.

In [5]:
df=lib.emailExtraction(df)
df

address  country hs_object_id  \
0             Blackpool  Rue, 6576  Ireland       416102   
1          Parkfield  Avenue, 5340  Ireland       413403   
2              Abourne   Lane, 876  Ireland       417951   
3     Chester      Crossroad, 7070  Ireland       419852   
4        Tilloch   Crossroad, 8332  Ireland       425352   
...                            ...      ...          ...   
6931             Cato   Hill, 5704  Ireland       463801   
6932        Howard Crossroad, 9998  England       454595   
6933              Walnut Rue, 3859  Ireland       454631   
6934         Kimberley  Pass, 3455  England       457622   
6935        Yorkshire  Alley, 5843  Ireland       461395   

                  industry          phone                          raw_email  \
0         Poultry and fish  0-774-386-624        zoe_owen450104633@acrit.org   
1     Fruit and vegetables  6-777-367-783  zara_rodwell1398442854@nickia.com   
2                  Milling  5-618-556-540  zara_freeburn1593147546@gmail.com   
3           Dairy products  1-161-604-327   winnie_walter538064895@sheye.org   
4                     Meat  5-645-416-200      zoe_owen1652446013@bungar.biz   
...                    ...            ...                                ...   
6931  Fruit and vegetables  2-301-310-740     rhea_grady231044837@deons.tech   
6932       Bakery products  6-747-016-018  hailey_farrell934796609@brety.org   
6933          Animal feeds  5-516-171-174  caleb_purvis1251615808@iatim.tech   
6934       Bakery products  0-605-727-343    gina_weasley83573127@naiker.biz   
6935  Fruit and vegetables  3-814-518-751    aeris_walsh769266811@gompie.com   

     technical_test___create_date       city          countryTuple  
0                      2021-07-13  Waterford  (Ireland, Waterford)  
1                      2021-01-09        NaN        (Ireland, nan)  
2                      2021-08-30        NaN        (Ireland, nan)  
3                      2021-02-10     Dublin     (Ireland, Dublin)  
4                      2021-11-02     Dublin     (Ireland, Dublin)  
...                           ...        ...                   ...  
6931                   2021-04-24  Waterford  (Ireland, Waterford)  
6932                   2021-12-04     London     (England, London)  
6933                   2021-02-23     Dublin     (Ireland, Dublin)  
6934                   2021-10-02   Plymouth   (England, Plymouth)  
6935                   2021-05-06       Cork       (Ireland, Cork)  

[6936 rows x 9 columns]

#### 2.3. Fix the phone numbers
For this part the phone numbers are going to be arange in certain format *(+XXX) XXXX XXXXXX*

In [6]:
df=lib.fixPhoneNumber(df)
df

address  country hs_object_id  \
0             Blackpool  Rue, 6576  Ireland       416102   
1          Parkfield  Avenue, 5340  Ireland       413403   
2              Abourne   Lane, 876  Ireland       417951   
3     Chester      Crossroad, 7070  Ireland       419852   
4        Tilloch   Crossroad, 8332  Ireland       425352   
...                            ...      ...          ...   
6931             Cato   Hill, 5704  Ireland       463801   
6932        Howard Crossroad, 9998  England       454595   
6933              Walnut Rue, 3859  Ireland       454631   
6934         Kimberley  Pass, 3455  England       457622   
6935        Yorkshire  Alley, 5843  Ireland       461395   

                  industry              phone  \
0         Poultry and fish   (+44) 7743 86624   
1     Fruit and vegetables  (+44) 6777 367783   
2                  Milling  (+44) 5618 556540   
3           Dairy products  (+44) 1161 604327   
4                     Meat  (+44) 5645 416200   
...                    ...                ...   
6931  Fruit and vegetables  (+44) 2301 310740   
6932       Bakery products  (+44) 6747 016018   
6933          Animal feeds  (+44) 5516 171174   
6934       Bakery products   (+44) 6057 27343   
6935  Fruit and vegetables  (+44) 3814 518751   

                              raw_email technical_test___create_date  \
0           zoe_owen450104633@acrit.org                   2021-07-13   
1     zara_rodwell1398442854@nickia.com                   2021-01-09   
2     zara_freeburn1593147546@gmail.com                   2021-08-30   
3      winnie_walter538064895@sheye.org                   2021-02-10   
4         zoe_owen1652446013@bungar.biz                   2021-11-02   
...                                 ...                          ...   
6931     rhea_grady231044837@deons.tech                   2021-04-24   
6932  hailey_farrell934796609@brety.org                   2021-12-04   
6933  caleb_purvis1251615808@iatim.tech                   2021-02-23   
6934    gina_weasley83573127@naiker.biz                   2021-10-02   
6935    aeris_walsh769266811@gompie.com                   2021-05-06   

           city          countryTuple  
0     Waterford  (Ireland, Waterford)  
1           NaN        (Ireland, nan)  
2           NaN        (Ireland, nan)  
3        Dublin     (Ireland, Dublin)  
4        Dublin     (Ireland, Dublin)  
...         ...                   ...  
6931  Waterford  (Ireland, Waterford)  
6932     London     (England, London)  
6933     Dublin     (Ireland, Dublin)  
6934   Plymouth   (England, Plymouth)  
6935       Cork       (Ireland, Cork)  

[6936 rows x 9 columns]

- Now, we proceed to drop the contacts with no email, because according to the [CRM Contact Hubspot](https://developers.hubspot.com/docs/api/crm/contacts) "*It is recommended to always include email, because email address is the primary unique identifier  to avoid duplicate contacts in HubSpot.*", so, the best option is to drop them. In a real-life situation, the best options would be to work with the client to evaluate the best way to proceed.

In [7]:
df.dropna(subset=['raw_email'], inplace=True)
df

address  country hs_object_id  \
0             Blackpool  Rue, 6576  Ireland       416102   
1          Parkfield  Avenue, 5340  Ireland       413403   
2              Abourne   Lane, 876  Ireland       417951   
3     Chester      Crossroad, 7070  Ireland       419852   
4        Tilloch   Crossroad, 8332  Ireland       425352   
...                            ...      ...          ...   
6931             Cato   Hill, 5704  Ireland       463801   
6932        Howard Crossroad, 9998  England       454595   
6933              Walnut Rue, 3859  Ireland       454631   
6934         Kimberley  Pass, 3455  England       457622   
6935        Yorkshire  Alley, 5843  Ireland       461395   

                  industry              phone  \
0         Poultry and fish   (+44) 7743 86624   
1     Fruit and vegetables  (+44) 6777 367783   
2                  Milling  (+44) 5618 556540   
3           Dairy products  (+44) 1161 604327   
4                     Meat  (+44) 5645 416200   
...                    ...                ...   
6931  Fruit and vegetables  (+44) 2301 310740   
6932       Bakery products  (+44) 6747 016018   
6933          Animal feeds  (+44) 5516 171174   
6934       Bakery products   (+44) 6057 27343   
6935  Fruit and vegetables  (+44) 3814 518751   

                              raw_email technical_test___create_date  \
0           zoe_owen450104633@acrit.org                   2021-07-13   
1     zara_rodwell1398442854@nickia.com                   2021-01-09   
2     zara_freeburn1593147546@gmail.com                   2021-08-30   
3      winnie_walter538064895@sheye.org                   2021-02-10   
4         zoe_owen1652446013@bungar.biz                   2021-11-02   
...                                 ...                          ...   
6931     rhea_grady231044837@deons.tech                   2021-04-24   
6932  hailey_farrell934796609@brety.org                   2021-12-04   
6933  caleb_purvis1251615808@iatim.tech                   2021-02-23   
6934    gina_weasley83573127@naiker.biz                   2021-10-02   
6935    aeris_walsh769266811@gompie.com                   2021-05-06   

           city          countryTuple  
0     Waterford  (Ireland, Waterford)  
1           NaN        (Ireland, nan)  
2           NaN        (Ireland, nan)  
3        Dublin     (Ireland, Dublin)  
4        Dublin     (Ireland, Dublin)  
...         ...                   ...  
6931  Waterford  (Ireland, Waterford)  
6932     London     (England, London)  
6933     Dublin     (Ireland, Dublin)  
6934   Plymouth   (England, Plymouth)  
6935       Cork       (Ireland, Cork)  

[6876 rows x 9 columns]

## 3. Managing duplicates
In this section, we will manage duplicates in the dataset using the email as the unique identifier, and the following rules:
- If the email is duplicated, we will keep the last entry.
- If the record duplicated has no information in any field (except the email), we will fill the missing information with the information of the previous record.
- For the industry field, we will concatenate the values of the duplicated records with a ; as a separator.

In [8]:
dfDuplicates=lib.duplicatesManage(df)
dfDuplicates.to_csv('dfDuplicates.csv', index=False)
dfDuplicates

address  country hs_object_id          industry  \
0          Besson  Avenue, 1005  Ireland       457508  Poultry and fish   
1         Sheraton   Hill, 1513  England       448185  Poultry and fish   
2             Edwin   Rue, 3956  England       454197           Milling   
3        Parkgate  Street, 5955  Ireland       460181   Bakery products   
4           Timothy  Lane, 9113  England       459017      Animal feeds   
...                         ...      ...          ...               ...   
6642  Tilloch   Crossroad, 8332  Ireland       425352              Meat   
6643       Blackpool  Rue, 6576  Ireland       416102  Poultry and fish   
6644     Catherine  Drive, 8284  Ireland       459018           Milling   
6645  Bermondsey    Drive, 6035  Ireland       460612  Poultry and fish   
6646        Yoakley Route, 2959  England       451190   Bakery products   

                  phone                            raw_email  \
0     (+44) 8532 182466       abbey_porter75675783@vetan.org   
1     (+44) 4868 808871   abdul_fisher1174469572@grannar.com   
2     (+44) 4231 262651     abdul_fisher666937132@twipet.com   
3     (+44) 5147 107006       abdul_fox1689938287@mafthy.com   
4     (+44) 1753 441164        abdul_fox1778689671@tonsy.org   
...                 ...                                  ...   
6642  (+44) 5645 416200        zoe_owen1652446013@bungar.biz   
6643   (+44) 7743 86624          zoe_owen450104633@acrit.org   
6644  (+44) 6706 675718  zoe_summers84316065@infotech44.tech   
6645  (+44) 7385 000772      zoe_wallace901459604@sveldo.biz   
6646  (+44) 8315 164044        zoe_weldon719495134@vetan.org   

     technical_test___create_date        city           countryTuple  
0                      2021-05-12    Limerick    (Ireland, Limerick)  
1                      2021-01-10      Oxford      (England, Oxford)  
2                      2021-12-07         NaN         (England, nan)  
3                      2021-07-12      Dublin      (Ireland, Dublin)  
4                      2021-03-16    Plymouth    (England, Plymouth)  
...                           ...         ...                    ...  
6642                   2021-11-02      Dublin      (Ireland, Dublin)  
6643                   2021-07-13   Waterford   (Ireland, Waterford)  
6644                   2021-03-15         NaN         (Ireland, nan)  
6645                   2021-05-29   Waterford   (Ireland, Waterford)  
6646                   2021-02-10  Winchester  (England, Winchester)  

[6647 rows x 9 columns]

## 4. Graphing the data.

In this section, we are going to explore the visualization of data through interactive graphics using the Plotly library in Python. We will focus on representing the number of people registered based on different aspects, such as countries, cities and dates. We will use bar, pie and line charts to visualize the distribution of people registered by country and the temporal evolution of the records. Something important is that with Plotly we can pass the cursor over some of the data and it will shows us the value of this, which allows us to have a better understanding of the data.

### 4.1. Country distribution
- First of all, we are going to check the country's distribution of the contacts. 

In [ ]:
# Group by country and count the number of contacts
CountryCount = dfDuplicates['country'].value_counts()

# Get the index and the values of the CountryCount
countryIndex = CountryCount.index.tolist()
countryValues = CountryCount.values.tolist()


In [ ]:
fig = go.Figure(data=go.Pie(labels=countryIndex, values=countryValues))
# Personalizar el gráfico
fig.update_layout(
    title="Contacts by Country",
)

# Mostrar el gráfico interactivo
fig.show()

- We can see that the country with more contacts is the England, followed by Ireland, but the difference is just around a 5%.

In [ ]:
# Funtion to get the colors for the chart
def colorsDefinition(listIndex):
    numColors = len(listIndex)
    colors = px.colors.qualitative.Plotly * (numColors // len(px.colors.qualitative.Plotly) + 1)
    # Return the colors for the chart
    return colors[:numColors]

### 4.2. City distribution
- Now we are going to see the distribution of the contacts by city. We are going to use the same function that we used for the country distribution, but this time we are going to group by city.

In [ ]:
# Group by city and count the number of contacts
cityCount = dfDuplicates['city'].value_counts()

# Get the index and the values of the cityCount
cityIndex = cityCount.index.tolist()
cityValues = cityCount.values.tolist()


In [ ]:
# Create the bar chart
fig = go.Figure(data=go.Bar(x=cityIndex, y=cityValues, marker=dict(color=colorsDefinition(cityIndex))))

# Personalize the graph
fig.update_layout(
    title="Distribution of People by City",
    xaxis_title="City",
    yaxis_title="Number of People"
)

# Show the interactive graph    
fig.show()

- Waterford is the city with more contacts, followed by Limerick and Dublin, but again the difference is not big.



### 4.3. Date of creation distribution
- Last, we are going to analyze the distribution of the contacts by the date of creation. We are going to use a Line chart to show the distribution of the contacts by the date of creation grouped by month.

In [ ]:
# Convert the column to datetime
dfDuplicates['technical_test___create_date'] = pd.to_datetime(dfDuplicates['technical_test___create_date'])

# Group by month and count the number of people
dateGroups = dfDuplicates.groupby(dfDuplicates['technical_test___create_date'].dt.month).size()

In [ ]:
# Get the months and the number of people
month = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October','November', 'December']
monthValues = dateGroups.values.tolist()

In [ ]:
# Create the Scatter plot
fig = go.Figure(data=go.Scatter(x=month, y=monthValues, marker=dict(color=colorsDefinition(month))))

# Personalize the graph
fig.update_layout(
    title="Number of People by Month",
    xaxis_title="Month",
    yaxis_title="Number of People"
)

# Show the interactive graph
fig.show()

- We notice that in this case March is the month with the most contacts created, and the month with the least contacts created is February, and from April the contacts created are not less than 540, so we can say that in this case the contacts created seem to be stable from April to December.

## 5. Importing the data into HubsSpot.
Finally, we will import the data into HubSpot. In this section, the `contactExtraction` function, will import contact by contact into HubSpot and it takes about 40 seconds to import 100 contacts. So, for this example, it will take about 40 minutes to import all the contacts. You will receive the next output message *'Contacts uploaded successfully'* once the import is finished. 

In [9]:
lib.importContacts(dfDuplicates,api_key_david)

'Contacts uploaded successfully'

> A point of improvement is instead of importing contact by contact, import the whole csv file at once, but I couldn't complete this part because I received  different errors in the request and I couldn't find the solution for them, if you could give me some feedback about this I would appreciate it.


## Optional Point

The document suggested an optional item related to storing the Street Address (in this case, Address) field in Hubspot. I completed this process using the property already existing in Hubspot called "Street Address" and the column "Address" from the dataframe. The process was done using the Import API of Hubspot, and you can see that the way to avoid losing the data was to take the most recent record for each duplicated contact, you could take a look at this process in the module `duplicatesManagment`.

---

## Questions Table
| Questions                                                                                                     | Answers                                                                                                                                                                                                                                                                                                                                                                     |
|---------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| What is your usual IDE?                                                                                       | I'm not really using an IDE, I use a code editor, Visual Studio Code, but I also have some experience with Spider and IDEs for different languages.                                                                                                                                                                                                                         |
| What are the advantages of this IDE over the others?                                                          | The advantages that I have by using VsCode are huge, first is the ability to personalize the environment that I use to develop, such as using extensions, create keyboard shortcuts, even easily configure the indent options, in this case 4 spaces by default for Python, another advantage is the integration with Git to easily manage the control version among other advantages. |
| Which of the items resulted in the most computational time for you? Add time in seconds.                      | The item that resulted in the most computing time was the import of the contacts in Hubspot, it took around 2600 seconds, for the 6647 contacts, but as I mentioned before I tried to upload the whole file but it give me some errors in the request that I couldn't solve, another option would be to import the contacts directly through the Hubspot CRM.               |
| If you have any public portfolio (e.g., IA, computer visión, data processing…), please share the link with us | I don't have much of a portfolio, but you can check out some of the projects I've been developing on my [Github](https://github.com/dhduartea) account.                                                                                                                                                                                                     |

---

I sinceraly thank you for the opportunity to participate in this evaluation test for the position of Data Engineer & Analyst at OTF. It has been an exciting and enriching challenge to demonstrate my skills in data handling, Python programming and integration with the HubSpot API. I hope I have showed some of my ability to face the challenges related to data processing, transformation and loading. I am excited about the possibility of being part of your team  and contributing to your success as a Data Engineer & Analyst.
    
Best regards,

    David Horacio Duarte Angulo